### ISSUE TITLE
Prevent havoc from patches that changed the scaling on the prefab.

### ISSUE THREAD

**Issue by Lisias** [net-lisias-ksp/TweakScale#137]
That's the history:

TweakScale depends of two thingies in order to know its way on the scaling stunt: the target scale, and the default scale of the part. with these two values, it can infer how much it should scale up or down a part (as well all the exponents).

**HOWEVER**, if any one of these values is corrupted, TweakScale is doomed - and will scale things badly (and even sent them into oblivion, as really damaged parts can end up with scaling the thing to small or to zero).

Some of the possible outcomes were explained on issue #34, and images reproduced here for convenience:

![](https://user-images.githubusercontent.com/64334/55228666-75687f80-51f9-11e9-8246-e37c25e1f67d.png)

![](https://user-images.githubusercontent.com/64334/54075134-4efa9880-427a-11e9-9d24-0917306434e3.png)

However, things are not that bad as it appears - it's possible to recover the correct scaling when we have **three** values: the current scale for the part, the "old" default scale in which the part was originally scaled, and the new current default scale that can be fetched from the prefab. And, in fact, TweakScale **was doing it** in the past, as this piece of code clearly demonstrates:

```
else if (defaultScale != scaleType.DefaultScale)
            {
                Log.warn("defaultScale has changed for part {0}: keeping relative scale.", part.name);
                currentScale *= scaleType.DefaultScale / defaultScale;
                defaultScale = scaleType.DefaultScale;
            }
```
([Source](https://github.com/net-lisias-ksp/TweakScale/blob/96e45e92cde40b7fa10baf84a90797d6927e68c6/Source/Scale/Scale.cs#L227))

Problem: it's not working nowadays. =/

By the time these problems started to bite me, I concluded that it was due bad patching. I was not wrong, I just wasn't right enough: TweakScale stopped handling this specific problem due external causes, and now I finally figured out why:

Between KSP 1.5.0 and 1.7.3 (I jumped directly from 1.4.5 to 1.7.3, so I don't know exactly when), KSP started to "migrate" crafts and savegames from older formats to the current one. The name of the stunt is [SaveUpgradePipeline](https://kerbalspaceprogram.com/api/class_save_upgrade_pipeline_1_1v180___module_aero_surface.html), and while doing this, someone decided it would be a good idea to automatically "fix" the Add'Ons too - besides not having enough knowledge to do it for the most complicated ones - and so things were leveled to the bottom: the most simple and/or common cases were tackled down, and the complicated ones were just ignored.

TweakScale was one of these ignored complicated cases.

Now that the UpgradePipeline is shoving prefab's current values indiscriminately into the part, the present TweakScale code lost the ability to detect this problem to fix it. And since I didn't understood until the moment how in hell this Pipeline stunt works (not to mention the myriad of constant problems due the terrible patching habits, disguised as "best practices" that still haunt TweakScale to this day), the sad true is that I didn't managed neither to prevent, neither to workaround the problem.

Until now.

The first hint I got was some new weird messages happening on loading crafts from different KSP versions:

```
[WRN 06:49:54.231] [Part]: PartModule indexing mismatch at FSfloatEndTail, index 2.
Node 'TweakScalerFSbuoyancy' found in loaded data, but 'Driftless' is defined in prefab.
Looking for TweakScalerFSbuoyancy in other indices...
[WRN 06:49:54.231] ...TweakScalerFSbuoyancy module found at index 3.
[WRN 06:49:54.231] [Part]: PartModule indexing mismatch at FSfloatEndTail, index 3.
Node 'TweakScale' found in loaded data, but 'TweakScalerFSbuoyancy' is defined in prefab.
Looking for TweakScale in other indices...
[WRN 06:49:54.231] ...TweakScale module found at index 4.
[LOG 06:49:54.232] [TweakScale] TRACE: OnLoad FSfloatEndTail True
```

This is clearly something trying to find its way out on a node from the craft file that is not in the same order than on prefab's one. And my debugging code tells me that this stunt is happening immediately before `OnLoad` is called on some code that identifies itself as `PART`. Humm... Interesting!

Dumping the ConfigNode on my `OnLoad` handler **before** and **after** calling `base.OnLoad` revealed that no change is being issued on this handler. But dumping the ConfigNode after it revealed the `defaultScale` mangled!!

**Oooooh, that's a bingo!!** :)
![](https://media.tenor.com/images/9e5675fd1e831b4d39c49845bdec2db0/tenor.gif)

Things are manageable now. **I have a chance to act before the UpgradePipeline do something**.

The fix is obvious now. Move the fixing code to `OnLoad`.

**Comment by Lisias** [comment]
Fixed on commit: https://github.com/net-lisias-ksp/TweakScale/commit/be333b516a047b6c1d7aabba17192b7144b807f6

**Comment by 7ranceaddic7** [comment]
By the gods, I wish I understood what I just read. 🤣

All I got out of that was Squad changed how the did something so they did something else that accounted for some simple modding but not the complex mods to make it easier for them to do what they wanted to do without breaking everything.

**Comment by Lisias** [comment]
It's simple, once you understand how things works on KSP. Really. :)

On the 1.3 to 1.4 era, they had a problem. Loading crafts and savegames on 1.4 from the 1.3 era was problematic - they were changing some things internally (for good reasons). So they needed a way to migrate these artifacts seamlessly - otherwise some people would not migrate to 1.4, and they would loose an opportunity to sell them the DLCs. (I would do the same, by the way - these guys need to eat!).

The problem started to happen when, somewhere between 1.5 and 1.7, they decided to migrate third parties Add'Ons too. It's a lost fight, it's **plain impossible** to guess the right things to be done for every fscking AddOn on the face of Kerbin, damnit!

That magnificent bork on Resources on KSP 1.9? It was due this stunt. Only **two** Add'Ons that mangle Resources survived that stunt, IFS (because the Author gave the finger to Squad some time ago and started to re-implement himself everything he needs on the Add'On, so the migration code just dry ran on him), and some other one that... "interestingly"... was already ready for the change when it happened (assuming the change were not made to exactly cope with his needs, of course).

Everybody else were plain royally screwed up by the stunt. TweakScale included.

I took a whole month to be aware of the stunt while helping a fellow Author to include TweakScale support on his Add'On: he said I didn't did by job right because the resources were not being updated. But I did the job right, and sent him screenshots of the thing working, and then we realized that he was working on KSP 1.9.0 while I was still developing on 1.7 **and** 1.8.

It took me **TWO WEEKS** until I finally realized exactly what was happening and implement a fix that would not break everybody else (my work colleagues helped  a lot on it, trying to lift or postpone any professional demands, so I could focus on this, by the way - nobody hinted me about this, I **really** dived deeply on KSP to solve this one). And almost an additional month more until I could figure out a way to allow the fix to work for anybody, and not just for people supporting TweakScale or KSP Recall directly.
